### ISSUE TITLE
Free energy analysis scripts MBAR/TI

### ISSUE THREAD

**Issue by ppxasjsm** [michellab/Sire#56]
Julien and I have discussed that the default free energy analysis method should from now on be MBAR and not TI any more. I have rewritten the code to now generate output files that can be used for MBAR analysis (as well as simple TI using trapezium integration)

These output files are ascii data (preferred format for easy inspection, by both Julien and myself.)
In order to now analyse this new data format with mbar I have started sketching a free_energy_analysis_mbar file. It works fine, but still needs a lot of sanity checking/commenting.  Currently free_energy_analysis is still supported and at least gradients.s3 files are written out. 

So far I didn't want to interfer with the old way of analysis too much, but at some point there should be a uniform free energy analysis tool. 

Are you particularly attached to the free_energy_analysis script? Do you have any ideas how to integrate this well. Should there be a backward compatibility, i.e. do we still want to support free_energy_analysis? I would like some thoughts/input on this before I make any rigorous decisions.

**Comment by chryswoods** [comment]
By default, waterswap and ligandswap calculate the free energy using FEP, TI and Bennetts simultaneously (is the MBAR?). analyse_freenrg analyses all three at the same time, giving the results for all three (with TI analysed and presented using trapezium rule and a quadrature method). It is very useful to have all methods used and available and simultaneously presented, so I would not be comfortable getting rid of TI or FEP and only leaving MBAR. I am very attached to analyse_freenrg.py and the code in corelib/src/libs/SireAnalysis. Any new analysis code should be added here and be in the same style and compatible with what is presented in corelib/src/libs/SireAnalysis.

There is definitely a tension between the different ways that Julien and I analyse free energies, and analyse_freenrg.py does try to support both ways. It does this by loading up your text-based gradient files into a Sire TI() object, that is then parsed and processed by the code as like any other freenrgs.s3 file.

I use a freenrgs.s3 file because waterswap and ligandswap use replica exchange, and the code handles all of the free energy data as a single unit. The free energy data is kept in the Sire binary format as I have found that it is easier for people to view the analysis output from each iteration of waterswap/ligandswap then to visualise the individual gradient data (there may be 1000 iterations of waterswap/ligandswap, so it is even too much data when I give a PMF for each of the 1000 iterations. To multiply this by 48 for 24 lambda values * 2 directions = 48000 gradients, plus 48000 FEP deltas plus 48000 BAR ratios is way too much.

Instead, the freenrgs.s3 file holds all of these in a compressed and easy-to-query format, with analyse_freenrg providing a tool that can extract data from this file with ease.

TL;DR I am very attached to analyse_freenrg.py, it is a much newer way of analysing free energies than you think, and changing it will significantly break waterswap and ligandswap.

**Comment by jmichel80** [comment]
My view is that it is useful to write the perturbed energies and gradients as text files to facilitate analysis by third party software, but that's not a big difference over an s3 file. I am keen to collect many datapoints per simulation as this enables computation of autocorrelation functions and subsampling to estimate errors with statistically independent samples. I prefer to use MBAR via a supported third-party tool rather than code and support multi-state BAR. I prefer to use a single default free energy estimator to simplify the output for lay users, but also output free energies from other estimators so that comparison can be made by advanced users.

Clearly we don't want to break other sire apps that depend on analyse_freenrg.py. It would be confusing to the users to have two free energy analysis scripts. 

I suggest as a resolution to extend the UI of analyse_freenrg.py so Chris's code is used for TI (via Ytreberg's regress),  FEP, BAR (via Chris's Sire code) and Antonia's code is used for MBAR (via PyMBAR).

**Comment by ppxasjsm** [comment]
Ok I think we should discuss this in Jülich/or skype in detail. I will not touch analyse_free_energy and just keep the analyse_free_energy_mbar as an extra analysis tool for now. Maybe we can find a good way to implement everything. I have very little idea of the capabilities of the Analysis toolkit and am happy to try and work with it if it is easy enough. 

BAR and MBAR are not the same and I would dread implementing MBAR in c++ if there is a perfectly working versions out there, that already catch most of the possible numerical horriblenesses that can happen when working with exponentials. 

On a different note, I am still not particularly convinced that the polynomial regression fits are the way to go for TI, cubic splines seem a lot better in my opinion, but I haven't tested this and the Ytrebergs paper failed to convince me of its superiority. 

All of this requires some detailed thought. What I have implemented now is a reasonable solution for now, but if we are to move towards say Hamiltonian replica exchange simulations between lambdas, this will require some clever thought. I am hesitant to extend analysie_freerg.py at the moment in an ad hoc way and maybe instead use indeed some of the time in Jülich to come up with a god data structure/concept for incorporating everything.

**Comment by chryswoods** [comment]
Ok - I think that it is a good idea that we spend some time in Juelich looking through this. From the sounds of it, it looks like I need to understand what you are doing and then see how that can be implemented into SireAnalysis. SireAnalysis already handles everything to do with replica exchange, and  I have RETI, REFEP and REBAR all working for waterswap, ligandswap and quantomm. I think it will be easier to implement MBAR into SireAnalysis as you need the C++ objects so that the replicas can hold binary-serialised copies of the data within their data structure. The RepExMove object then knows how to exchange replicas while keeping the free energy data with its proper lambda value.

I think that the best thing would be for us to spend about 1-2 hours talking about this. First, you can show me MBAR and how you have implemented that, and then I will show you SireAnalysis, and how this integrates with the replica exchange code.

**Comment by ppxasjsm** [comment]
I am not using my own implementation of MBAR, a basic MBAR code is written quite quickly, but there are a few ways to optimize things concerning numerical stability etc, and it is even faster to use an external pymbar module, from the Chodera lab, at least right now. I am happy to explore other possibilities, but I don't want to spend ages (and I always take alot longer than I want to go from an initial working prototype to a well matured code) reimplementing readily available software.

**Comment by jmichel80** [comment]
We shouldn't implement stuff that is already debugged by others. The actual
data crunching should be done via pymbar, and this can be linked from code
within SireAnalysis. A good example is how analyse_freenrg.py actually
re-uses the regress C code (IIRC this has been added somewhere in
SireAnalysis).
It would be great if the Sire's repex code can be efficiently used to do
all the book keeping needed to manage multiple somd simulations running
concurrently on different GPUs (that may not all be on the same node).

I think one of the difficulties we have to address is that Antonia and I
are more comfortable implementing new features via Python, whereas Chris
more naturally re-uses the C++ Sire's API.


--------------------------------------------------------------
Dr. Julien Michel,
Royal Society University Research Fellow
Room 263, School of Chemistry
Joseph Black Building, University of Edinburgh
David Brewster Road

Edinburgh

EH9 3FJ
United Kingdom
phone:   +44 (0)131 650 4797
http://www.julienmichel.net/
-------------------------------------------------------------

On Thu, Sep 3, 2015 at 10:45 AM, ppxasjsm <notifications@github.com> wrote:

> I am not using my own implementation of MBAR, a basic MBAR code is written
> quite quickly, but there are a few ways to optimize things concerning
> numerical stability etc, and it is even faster to use an external pymbar
> module, from the Chodera lab, at least right now. I am happy to explore
> other possibilities, but I don't want to spend ages (and I always take alot
> longer than I want to go from an initial working prototype to a well
> matured code) reimplementing readily available software.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/michellab/Sire/issues/56#issuecomment-137393843>.
>

The University of Edinburgh is a charitable body, registered in
Scotland, with registration number SC005336.
